src/sd-jwt/index.ts

```diff
@@ -61,27 +70,64 @@ export const decode = <S extends z.AnyZodObject>(
  * @param claims The list of claims to be disclosed
  *
  * @throws {ClaimsNotFoundBetweenDislosures} When one or more claims does not relate to any discloure.
- * @returns The encoded token with only the requested disclosures
+ * @throws {ClaimsNotFoundInToken} When one or more claims are not contained in the SD-JWT token.
+ * @returns The encoded token with only the requested disclosures, along with the path each claim can be found on the SD-JWT token
  *
  */
-export const disclose = (token: string, claims: string[]): string => {
+export const disclose = async (
+  token: string,
+  claims: string[]
+): Promise<{ token: string; paths: { claim: string; path: string }[] }> => {
   const [rawSdJwt, ...rawDisclosures] = token.split("~");
+  const { sdJwt, disclosures } = decode(token, SdJwt4VC);
 
   // check every claim represents a known disclosure
   const unknownClaims = claims.filter(
     (claim) =>
-      !rawDisclosures.map(decodeDisclosure).find(([, name]) => name === claim)
+      !rawDisclosures
+        .map(decodeDisclosure)
+        .find(({ decoded: [, name] }) => name === claim)
   );
   if (unknownClaims.length) {
     throw new ClaimsNotFoundBetweenDislosures(unknownClaims);
   }
 
   const filteredDisclosures = rawDisclosures.filter((d) => {
-    const [, name] = decodeDisclosure(d);
+    const {
+      decoded: [, name],
+    } = decodeDisclosure(d);
     return claims.includes(name);
   });
 
-  return [rawSdJwt, ...filteredDisclosures].join("~");
+  // compose the final disclosed token
+  const disclosedToken = [rawSdJwt, ...filteredDisclosures].join("~");
+
+  // for each claim, return the path on which they are located in the SD-JWT token
+  const paths = await Promise.all(
+    claims.map(async (claim) => {
+      const disclosure = disclosures.find(
+        ({ decoded: [, name] }) => name === claim
+      );
+      const hash = await sha256ToBase64(disclosure?.encoded || "");
+
+      // _sd is defined in verified_claims.claims and verified_claims.verification
+      // we must look into both
+      if (sdJwt.payload.verified_claims.claims._sd.includes(hash)) {
```

Comment:
I think I didn't understand. This is not a fallback, we have to search in both sets - claims and verification. The point is, as claims are bigger and are more likely to be hit, we perform the search sequentially (is it in a claim? If no, it must be in verification...)

Reply:
sorry, maybe I didn't explain myself well. I was referring to the empty string fall back option on line 111

Reply:
It's just impossible to happen (but I didn't bother to instruct the type system accordingly). 

It's very ugly but it has an advantage: it does not pollute the flow with an option impossible to happen. But I'll think on a more elegant solution.

Reply:
yes, if it's impossible to happen then better find a more elegant solution. Otherwise NP

Reply:
Solved in https://github.com/pagopa/io-react-native-wallet/pull/22/commits/682dc52f0542632ce7f04e32a84a44ce2ec4c262

I decide to do the claim presence check in the same iteration of the path sought. So that we are sure the disclosure exists and we save one loop.